### PR TITLE
Tyler Porter 

### PR DESCRIPTION
**Pull request by tylerpporter**

### CODE REVIEW

**Comment by tylerpporter** [comment]
I completed iteration 4 after the fact for my own peace of mind and to receive feedback on my methods and testing. I made a really stupid mistake during this independent challenge that cost me about 30 minutes (I forgot to call the add_show method in my test so my actors_by_show method kept returning an empty hash... ugh)

File: lib/character.rb [hunk 1]
```diff
@@ -1 +1,19 @@
...
+  end
+
+  def actor
+    @character_info[:actor]
+  end
+
+  def salary
+    @character_info[:salary].to_i
+  end
+
```
Comment (aziobrow) [inline lib/character.rb]: nice job catching that this class takes a single argument and that it's a hash, so accessing the values accordingly.  

You could have (and it would probably be preferable to, for the sake of readability) handled this all in initialize. 

```
def initialize(character_info)
    @name = character_info[:name] 
    @actor = character_info[:actor] 
    @salary = character_info[:salary]
end
```

Note that you don't need the `@character_info = character_info` if we handle it all in initialize, since we don't ever really care about the hash in its totality-- we only really need it to begin with to grab out the specific values we care about

File: lib/show.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  attr_reader :name, :creator, :characters
+
+  def initialize(name, creator, characters)
+    @name = name
+    @creator = creator
+    @characters = characters
+  end
+
+  def total_salary
+    @characters.sum {|character| character.salary}
```
Comment (aziobrow) [inline lib/show.rb]: great choice of enumerable

File: lib/show.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def highest_paid_actor
+    @characters.max_by {|character| character.salary}.actor
```
Comment (aziobrow) [inline lib/show.rb]: perfect

File: lib/show.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def actors
+    @characters.map {|character| character.actor}
```
Comment (aziobrow) [inline lib/show.rb]: beautiful

File: lib/network.rb [hunk 5]
```diff
@@ -1 +1,45 @@
...
+  def main_characters
+    main_characters = []
+    @shows.each do |show|
+      show.characters.each do |char|
+        if (char.name == char.name.upcase) && (char.salary > 500_000)
+          main_characters << char
+        end
+      end
+    end
+    main_characters
```
Comment (aziobrow) [inline lib/network.rb]: great job figuring out the boolean logic-- this gets you exactly what you need. 

You could have used `map` in the outer loop (line 15) to avoid creating that empty array and shoveling into it.

Comment (aziobrow) [reply lib/network.rb]: You're also using nested iteration (an enumerable inside the block of another enumerable), which is good to know how to execute, because sometimes it isn't avoidable. In this case though, it would be avoidable, by doing something like this:

```
  def main_characters
     network_characters = shows.flat_map {|show| show.characters}
     network_characters.find_all {|character| character.salary > 500000 && (character.name == character.name.upcase)}
end
```

The latter solution here is preferable a) because it's easier to read and b) because it is more efficient. Nested iteration causes us to loop an exponential number of times, whereas splitting them out causes us to loop a linear number of times. So if we have 3 characters for 3 shows, nested iteration loops over 9 elements (3 characters for the first show, 3 for the second, 3 for the third), whereas splitting them out iterates only 6 times (once for each of the 3 shows, and once for each of the 3 resulting characters). This obviously isn't a big deal with the size of the data here, but it becomes a very big deal when we're talking about large sets. This is the idea behind Big O Notation (https://en.wikipedia.org/wiki/Big_O_notation) which is a common enough interview question/topic that it's worth being aware of.

tl;dr-- avoid nested iteration when possible, although it isn't always possible.

File: lib/network.rb [hunk 6]
```diff
@@ -1 +1,45 @@
...
+  end
+
+  def actors_by_show
+    by_show = {}
+    @shows.each {|show| by_show[show] = show.actors}
+    by_show
```
Comment (aziobrow) [inline lib/network.rb]: Nice. This works well, and particularly nice job reusing the code you already wrote to get `show.actors`

You could also have used reduce, as reduce is usually the best tool for creating a hash 

```
 def actors_by_show
    @shows.reduce({}) do |result, show|
           result[show] = show.actors
           result
     end
end 
```

File: lib/network.rb [hunk 7]
```diff
@@ -1 +1,45 @@
...
+  end
+
+  def shows_by_actor
+    by_actor = {}
+    @shows.each do |show|
+      show.actors.each do |actor|
+        by_actor[actor].nil? ? by_actor[actor] = [show] : by_actor[actor] << show
```
Comment (aziobrow) [inline lib/network.rb]: this is super solid, although probably pushing on the upper bounds of the amount of logic/work that we'd ideally like to see contained within a single ternary 

Again, reduce is a good tool here:

```
def shows_by_actor
    all_characters.reduce({}) do |result, character|
        result[character.actor] = @shows.find_all {|show| show.characters.include?(character)}
    end
end
```

File: lib/network.rb [hunk 8]
```diff
@@ -1 +1,45 @@
...
+      end
+    end
+    by_actor
+  end
+
+  def prolific_actors
+    shows_by_actor.select {|actor, show| actor if show.size > 1}.keys
```
Comment (aziobrow) [inline lib/network.rb]: remember that you can't tell a `select`/`find_all` what to return-- if you find yourself trying to do that (as you are here, when you say `actor if show.size > 1`, you probably want a different enumerable.  In this case, map would do it-- swap the `select` for a map and replace the `keys` with a `compact` 

```
def prolific_actors 
   shows_by_actor.map {|actor, shows| actor if shows.size > 1}.compact
end 
```

In this case, it's obviously not saving you any lines of code, but it avoids the misuse of `select`.

File: test/character_test.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    assert_instance_of Character, @kitt
+  end
+
+  def test_it_can_return_character_info
+
+    assert_equal "KITT", @kitt.name
+    assert_equal "William Daniels", @kitt.actor
+    assert_equal 1000000, @kitt.salary 
+  end
+
```
Comment (aziobrow) [inline test/character_test.rb]: 👍

File: test/network_test.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    assert_equal "NBC", @nbc.name
+    assert_equal [], @nbc.shows
+  end
+
+  def test_it_can_add_shows
+    @nbc.add_show(@knight_rider)
+    @nbc.add_show(@parks_and_rec)
+
+    assert_equal 2, @nbc.shows.size
+    assert_equal Show, @nbc.shows[0].class
```
Comment (aziobrow) [inline test/network_test.rb]: this is an okay test, but it's less code and more robust to just assert exactly what you're looking for. this would pass if somehow the wrong shows got in there. 

`assert_equal [@knight_rider, @parks_and_rec]. @nbc.shows`

If you're worried about order dependency, you can do 

`assert_includes @knight_rider, @nbc.shows`
`assert_includes @parks_and_rec, @nbc.shows`

File: test/network_test.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def test_it_can_return_main_characters
+    @nbc.add_show(@knight_rider)
+    @nbc.add_show(@parks_and_rec)
+
+    assert_equal Array, @nbc.main_characters.class
+    assert_equal "KITT", @nbc.main_characters[0].name
```
Comment (aziobrow) [inline test/network_test.rb]: again, this test is too loose, and you could do a better test in less lines.  it would pass if you had multiple characters in there-- all that it's checking is that it's an array, of which KITT is the first one.   Especially since this is only one value and there's no risk of erroneous order dependency, just: 

`assert_equal ["KITT"], @nbc.main_characters`

File: test/network_test.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  def test_it_can_sort_actors_by_show
+    @nbc.add_show(@knight_rider)
+    @nbc.add_show(@parks_and_rec)
+
+    assert_equal Hash, @nbc.actors_by_show.class
+    assert_equal Show, @nbc.actors_by_show.keys[0].class
+
+    actual = @nbc.actors_by_show.values.any? {|values| values == ["Amy Poehler", "Nick Offerman"]}
+
+    assert actual
```
Comment (aziobrow) [inline test/network_test.rb]: Why are you only checking for the parks and rec actors? What if the knight rider actors weren't in there?  What if only the first key was a show, and the other keys were somehow strings?  Just assert the exact value that you're expecting-- especially on something like this, where you just have to copy it out of the interaction pattern. 

As a reminder-- the lens for testing is not just that you know it's working. It's also as living documentation for future developers-- if I was coming into this code and was tasked with making updates to your `actors_by_show` method, i would really have very little idea from looking at this test what the expected output should be.  I'd have to carefully read and pry through your actual code to figure out what was supposed to be happening. Avoiding that is a big part of what tests are for.

File: test/network_test.rb [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    @nbc.add_show(@baywatch)
+    expected = {
+                "David Hasselhoff" => [@knight_rider, @baywatch],
+                "William Daniels" => [@knight_rider],
+                "Amy Poehler" => [@parks_and_rec],
+                "Nick Offerman" => [@parks_and_rec]
+                }
+
+
+    assert_equal expected, @nbc.shows_by_actor
```
Comment (aziobrow) [inline test/network_test.rb]: this is a good test

File: test/network_test.rb [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def test_it_can_return_prolific_actors
+    @nbc.add_show(@knight_rider)
+    @nbc.add_show(@parks_and_rec)
+    @nbc.add_show(@baywatch)
+
+    assert_equal ["David Hasselhoff"], @nbc.prolific_actors
```
Comment (aziobrow) [inline test/network_test.rb]: so is this
